solidity/contracts/bridge/Bridge.sol

```diff
@@ -141,36 +180,165 @@ contract Bridge {
         bytes memory txInputVector,
-        bytes memory txOutput,
+        bytes memory txOutputVector,
         bytes4 txLocktime,
         bytes memory merkleProof,
         uint256 txIndexInBlock,
         bytes memory bitcoinHeaders
     ) external {
-        // TODO We need to read `fundingTxHash`, `fundingOutputIndex` and
-        //      P2SH script depositor address from `txInputVector`.
-        //      We then hash them to obtain deposit identifier and read
-        //      DepositInfo. From DepositInfo we know what amount was declared
-        //      by the depositor in their reveal transaction and we use that
-        //      amount to update their Bank balance, minus fee.
-        //
-        // TODO We need to validate if the sum in the output minus the
-        //      amount from the previous wallet balance input minus fees is
-        //      equal to the amount by which Bank balances were increased.
-        //
-        // TODO We need to validate txOutput to see if the balance was not
-        //      transferred away from the wallet before increasing balances in
-        //      the bank.
-        //
-        // TODO Delete deposit from unswept mapping or mark it as swept
-        //      depending on the gas costs. Alternativly, do not allow to
-        //      use the same TX input vector twice. Sweep should be provable
-        //      only one time.
+        validateSweepProof(
+            txVersion,
+            txInputVector,
+            txOutputVector,
+            txLocktime,
+            merkleProof,
+            txIndexInBlock,
+            bitcoinHeaders
+        );
+        updateBalances(txInputVector, txOutputVector);
     }
 
     // TODO It is possible a malicious wallet can sweep deposits that can not
     //      be later proved on Ethereum. For example, a deposit with
     //      an incorrect amount revealed. We need to provide a function for honest
     //      depositors, next to sweep, to prove their swept balances on Ethereum
     //      selectively, based on deposits they have earlier received.
+
+    function updateBalances(
+        bytes memory txInputVector,
+        bytes memory txOutputVector
+    ) internal {
+        // We need to read `fundingTxHash`, `fundingOutputIndex` and
+        // P2SH script depositor address from `txInputVector`.
+        // We then hash them to obtain deposit identifier and read
+        // DepositInfo. From DepositInfo we know what amount was declared
+        // by the depositor in their reveal transaction and we use that
+        // amount to update their Bank balance, minus fee.
+        uint256 varIntDataLength;
+        uint256 inputCounter;
+        (varIntDataLength, inputCounter) = BTCUtils.parseVarInt(txInputVector);
+        uint256 offset = 1 + varIntDataLength;
+
+        for (uint256 i = 0; i < inputCounter; i++) {
+            bytes memory input =
+                txInputVector.slice(offset, txInputVector.length - offset);
```

Comment:
Why not just `txInputVector.extractInputAtIndex`?

Reply:
Done. 
Although `txInputVector.extractInputAtIndex` will not be very efficient in case there are multiple inputs.
For example we will be calculating the number of inputs inside this function all the time, while parsing the `txInputVector`.

Traversing  `txVectorInput` is done twice now - during validation and updating balances. We could try to optimise in the future.

Reply:
If not using `txInputVector.extractInputAtIndex` is cheaper we should definitely got that way.